### PR TITLE
Use Playwright to run WP Desktop tests

### PR DESCRIPTION
**Pull request by scinos**
## Changes

Uses Playwright to run WP Desktop E2E tests.

### Features

- Automatically capture session's HAR, console logs, and screenshots.
- Unified testing library with Calypso E2E. Eventually, it may be able to reuse `@automattic/calypso-e2e`.
- Run in Jest, like the rest of Calypso.

### Caveats

- We can't use `@automattic/calypso-e2e` right now, so I had to duplicate some logic for the Login flow. The reason is `desktop/` is actually a separate dependency tree with its own `yarn.lock`. As such, it doesn't see the monorepo package `@automattic/calypso-e2e`. First, we need to merge `desktop` into the main dependency tree (#55184).

### Pending tasks

To be done after the merge:

- [ ] Clean up `ci-desktop` docker image. We don't need to install Chrome anymore. In fact, we may be able to just use the same image E2E test is using (`ci-e2e`)

### CODE REVIEW

File: desktop/README.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ Refer to the [development guide](docs/development.md) for help with how the app
 ## Running The End-To-End Test Suite
 
 1. Set the environment variables `E2EUSERNAME` and `E2EPASSWORD`.
-2. Use `npm run e2e` or `make e2e` to invoke the test suite.
+2. Use `yarn run test:e2e` or `make e2e` to invoke the test suite.
```
Comment (nsakaimbo) [inline desktop/README.md:41]: 😆 Nice catch! 👍

**Comment by nsakaimbo** [comment]
Thanks for pushing ahead with this, @scinos!
 
> Electron support in Playwright is experimental. There are a few features (video recording, traces) that I couldn't get to work.

Video capture is nice, but not a deal breaker for this change so that's fine. It would be great if we could retain Electron logging at the very least, although worst case we can punt on that as an enhancement to revisit for later.

> Not sure if/how this will work for Mac and Windows tests.

We've been running e2e for Mac (Circle) and Linux (Circle and TC) until now, not Windows, so at the very least would be nice to maintain e2e for the same platforms (e2e for all platforms would be ideal but at least 2/3 gives us reasonable assurance that things are working for all). Perhaps Playwright might make it easier to run e2e on Windows as well in the future, although  I wouldn't consider Windows e2e a dealbreaker for this PR.

> They seem to be flakey. The screenshot shows that sometimes the main page is not loaded. Locally, I've seen how it opens the form but doesn't fill the input with the username. When the page fails to load, the reported URL is file:///calypso/desktop/release/linux-unpacked/resources/app/dist/public_desktop/index.html instead of a calypso.live, URL, but I don't know what that means.

The Desktop app's window comprises two elements - a "parent" `BrowserWindow` which hosts the "browser UI" of the Desktop app - the Back, Forward and Home navigation controls. These elements are local to the Electron app and are defined in the file `https://github.com/Automattic/wp-calypso/blob/trunk/desktop/public_desktop/index.html`. The parent `BrowserWindow` also hosts a "child" `BrowserView`, which is the object in which Calypso/WordPress are loaded. You can see how the `BrowserView` and `BrowserWindow` objects are initialized [here](https://github.com/Automattic/wp-calypso/blob/a58462164c39ae7dc0ad2620fafa83a64c91f3e7/desktop/app/mainWindow/index.js#L51-L62), and [this](https://github.com/Automattic/wp-calypso/blob/a58462164c39ae7dc0ad2620fafa83a64c91f3e7/desktop/app/mainWindow/index.js#L146) is the point at which the remote URL (wordpress.com) is loaded into the `BrowserView`. So perhaps the tests are failing because Playwright is looking at the `BrowserWindow` object, and not the `BrowserView`? Not sure but I'll try to poke around the changes to see if any ideas come to mind regarding the flakiness. 🤔

**Comment by nsakaimbo** [comment]
- I tried enabling Electron logging in 32bfde2a937d91d732747fb0d1fee863bb008e54. Setting the `WP_DEBUG_LOG` environment variable should do the trick (and is [how we do it today](https://github.com/Automattic/wp-calypso/blob/007551ca0a294563e47935087f58e0f8e534133a/desktop/e2e/run.js#L132-L134)), but it seems that Playwright isn't passing along the environment variable overrides when launching the app.

- I also just realized that we would want to override the `WP_DESKTOP_BASE_URL` env var as well ([example](https://github.com/Automattic/wp-calypso/blob/007551ca0a294563e47935087f58e0f8e534133a/desktop/e2e/run.js#L136-L137)), because this is how the application is pointed to the `calypso.live` URL for each PR.

- Regarding the first two points, it's really odd that I seem unable to override any env vars via the "launch" API provided by Playwright.

- Regarding my earlier suspicion of Playwright not pointing to the correct thing (i.e. `BrowserView` vs `BrowserWindow` object) - I don't think that's the case. In my testing locally I observed that - as you say - the test is flaky, so it works _some_times. Meaning that this shouldn't be a concern. I tried adding some delays to see if that would improve things, but no luck so far.

If I get some time later I'll try to play around with these changes some more, but just sharing what I've observed so far.

**Comment by dpasque** [comment]
I'm realizing now just how little I know about `electron` and our desktop applications 🙃 

That being said, I think this very much worth pushing forward on. I think having all of our E2E tests on the same platform, and able to ultimately re-use the same libraries, will be a huge boon to productivity and maintenance ease! I will try to get more familiar with the desktop app to hopefully be able to help with ironing out the kinks here!

File: desktop/test/e2e/specs/login.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
+/* eslint-disable no-console, import/no-nodejs-modules */
+const { createWriteStream } = require( 'fs' );
+const { mkdir } = require( 'fs/promises' );
+const path = require( 'path' );
+const { _electron: electron } = require( 'playwright' );
+
+// Linux:
+const APP_PATH = path.join( __dirname, '../../../release/linux-unpacked/wpcom' );
+// Mac:
+// const APP_PATH = path.join(
```
Comment (nsakaimbo) [inline desktop/test/e2e/specs/login.js]: If testing locally on Mac, you can replace the `APP_PATH` argument with this one.

File: desktop/test/e2e/specs/login.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+// 	'MacOS',
+// 	'WordPress.com'
+// );
+const CONSOLE_PATH = path.join( __dirname, '../results/console.log' );
+const SCREENSHOT_PATH = path.join( __dirname, '../results/screenshot.png' );
+const HAR_PATH = path.join( __dirname, '../results/network.har' );
+const VIDEO_PATH = path.join( __dirname, '../results/video.webm' );
+
+const timestamp = new Date().toJSON().replace( /:/g, '-' );
+const appLogPath = path.resolve( __dirname, '..', 'results', `app-${ timestamp }.log` );
```
Comment (nsakaimbo) [inline desktop/test/e2e/specs/login.js]: This file isn't being created, which implies that the environment variable `WP_DEBUG_LOG` is not being passed through to the app.

**Comment by nsakaimbo** [comment]
@scinos I _think_ I I'm getting closer to addressing the flakiness in 1f1fba8efee1a0b77c6795968c118e22ab22052f.   From my testing, waiting for all "frames" in the window to load ensures that Playwright will be able to find the elements it's looking for when the actual tests begin. The tests appear to be working more consistently locally and in CI, although there is still some intermittent flakiness as with the Circle Linux job:

https://app.circleci.com/pipelines/github/Automattic/wp-calypso/114964/workflows/20bb9fd8-38fa-4069-94ff-edd41620db88/jobs/1071970 

However when I run the tests locally (on Mac) it seems to be working 100% of the time.

**Comment by nsakaimbo** [comment]
Update: I think we're getting pretty close, and aside from progress on the flakiness a couple issues still remain - namely the fact that Playwright's "launch" API isn't overriding any environment variables:

- Not being able to set `WP_DESKTOP_BASE_URL` means the app isn't using `calypso.live` for each PR. This is definitely something we need to fix as we recently ran into issues where breaking changes were introduced due to the mis-match between the webapp being loaded by the Desktop app, and the changes introduced by a specific PR.

-  Since we're unable to set `WP_DEBUG_LOG` variable as well, we can't persist the application/Electron logs, which (especially in lieu of video capture) would be super helpful. Although I see that at least a screenshot is being persisted, so perhaps this point is less of a concern?

**Comment by nsakaimbo** [comment]
> I think having all of our E2E tests on the same platform, and able to ultimately re-use the same libraries, will be a huge boon to productivity and maintenance ease

@dpasque Good point! It would be a huge win if we could streamline things to re-use not only the same testing frameworks, but also be able to re-use the same _tests_ as well. It would indeed make maintenance much easier, which in turn would also as help us catch any inadvertent discrepancies in functionality a lot more proactively.

**Comment by scinos** [comment]
> also be able to re-use the same tests as well.

🤯 
That's really interesting! We could have the exact same suite running in `desktop`, `mobile` (like today) and just add a new environment, `app`

**Comment by scinos** [comment]
Regarding env vars, they seem to be passed to the app, at least when running locally on macOS:

image

**Comment by scinos** [comment]
I still get flaky tests in TeamCity. It looks like it has problems identifying the "first window":

* When the tests passes, `electronApp.firstWindow().url()` returns a calypso.live URL, as expected.
* When the test fails, `electronApp.firstWindow().url()` returns `.../desktop/release/linux-unpacked/resources/app/dist/public_desktop/index.html`

Looks like sometimes it detects `BrowserWindow` as the "first window", and some other times it detects the `BrowserView` object?

File: desktop/test/e2e/specs/login.js [hunk 4]
```diff
@@ -49,20 +51,36 @@ describe( 'User Can log in', () => {
...
 				WP_DEBUG_LOG, // This will override logging path from the Electron main process.
 				// Ensure other CI-specific overrides (such as disabling the auto-updater)
 				DEBUG: true,
 				CI: true,
 			},
 		} );
+
+		// Find main window. Playwright has problems identifying the main window when using `firstWindow`, so we
+		// iterate over all windows and find it by URL.
+		for ( const window of await electronApp.windows() ) {
```
Comment (nsakaimbo) [inline desktop/test/e2e/specs/login.js:60]: Interesting - does this change seem to make performance more consistent? From my understanding there is only a single window object, and the element in which WordPress/Calypso is loaded is essentially a "frame" in that object. You can verify this by examining the following:

```
for ( const [ _, frame ] of mainWindow.frames().entries() ) {
  console.log( 'Frame title: ', frame.title();
  console.log( 'Frame URL:', frame.url();
}
```

However it's also quite possible that perhaps Playwright sees as different "windows" as well, so doesn't hurt to try any and all ideas.

On a related note, it just occurred to me to check whether we're structuring the login test in the same way as in Calypso? For example, in my testing, _timing_ seemed to be the main issue - for some reason, Playwright was able to successfully wait for specific selectors during some trials, and failed to latch onto the same selectors during others. Even though these selectors are in fact always present when inspected via developer tools! Really odd. 🤔 The TL;DR of my question here is perhaps the Calypso login test waits/selects elements in a very specific way that avoids this problem? Maybe there's some other elements we're missing. Not sure but might be worth double-checking. (Although if Calypso isn't using Playwright at all for the same tests this point is moot, I suppose.)

Comment (scinos) [reply desktop/test/e2e/specs/login.js:60]: Looking at [Playwright code](https://github.com/microsoft/playwright/blob/master/packages/playwright-core/src/client/electron.ts#L89), main window is either literally the first window, or the first object that throws the event `window`. In that context, a "window" is an object that throws the event `page`.

I don't know if those events are Playwright specific or something more generic. But I guess that given it's based on events there is always the chance of race conditions, it sounds brittle to me.

Comment (scinos) [reply desktop/test/e2e/specs/login.js:60]: Calypso uses https://github.com/Automattic/wp-calypso/blob/74d1739c394880525967228e82bc835f5ba7cdbd/packages/calypso-e2e/src/lib/pages/login-page.ts#L102 to exercise the Login window. The login flow I implemented in this test is a super-simplified version, mostly to prove it works.

I'd love to use the same library used by Calypso, but for that to work first, we need to migrate Desktop to yarn v 3.1, and then integrate it with the main dependency tree so it has visibility on `@automattic/calypso-e2e`. First step of that journey is in https://github.com/Automattic/wp-calypso/pull/57487

Comment (scinos) [reply desktop/test/e2e/specs/login.js:60]: Re: stability. After this change I had 5 green runs in a row in TeamCity, so it seems to be helping!

Comment (scinos) [reply desktop/test/e2e/specs/login.js:60]: Re: Re: stability. I'm getting lots of green builds in CircleCI as well. I'm going to clean up the code a bit so it is ready to be merged.

**Comment by nsakaimbo** [comment]
oh, quick note: we'll need to persiste the logs in both Circle and TC so it will save the contents of the "results" folder instead.

Circle: persist logs [Mac](https://github.com/Automattic/wp-calypso/blob/94791f39a0d7166579f8eda09ff4b2475f2dc804/.circleci/config.yml#L308) and [Linux](https://github.com/Automattic/wp-calypso/blob/94791f39a0d7166579f8eda09ff4b2475f2dc804/.circleci/config.yml#L382).

**Comment by scinos** [comment]
> oh, quick note: we'll need to persiste the logs in both Circle and TC so it will save the contents of the "results" folder instead.

Done!

**Review by noahtallen — APPROVED** [review]
TC looks good to me! If I'm reading this right, are we removing our existing Desktop test suite and replacing it with a very minimal one that just tests login? And then plan to expand that in the future with the normal Calypso suite?

**Review by worldomonation — APPROVED** [review]
I think this looks good and is an excellent starting point for potential future expansion!

**Comment by nsakaimbo** [comment]
Looks good! I also confirmed that the Team City job is using the correct `calypso.live` endpoint for e2e tests ([link](https://container-lucid-babbage.calypso.live/log-in)).

> are we removing our existing Desktop test suite and replacing it with a very minimal one that just tests login? And then plan to expand that in the future with the normal Calypso suite?

@noahtallen The test suite for the Desktop app used to be a little more expansive, but was narrowed down to just the "boot up and login test" after the Desktop app was re-architected. A more extensive test suite would have been ideal, but other issues have always taken priority and only the bare minimum has been implemented so far.

@scinos's amazing work here does set the stage to perhaps re-use the existing Calypso tests for the Desktop app in the future, which would be an amazing win with respect to ongoing maintenance.
